src/scanner.ts

```diff
@@ -1258,46 +1202,57 @@ export class Scanner {
         return this.scanPunctuator();
     };
 
-    start() {
-        this.scanning = true;
-
-        this.skipComment();
-
-        this.startIndex = this.lastIndex = this.index;
-        this.startLineNumber = this.lastLineNumber = this.lineNumber;
-        this.startLineStart = this.lastLineStart = this.lineStart;
-
-        this.lookahead = this.advance();
-        if (this.addToken) {
-            this.addToken(this.lookahead);
-        }
-
-        this.scanning = false;
-    };
-
-
-    lex() {
-        this.scanning = true;
+    advance() {
+        let token;
+        if (this.source[this.index] !== '/') {
+            token = this.lex();
+        } else {
+            // Using the following algorithm:
+            // https://github.com/mozilla/sweet.js/wiki/design
 
-        this.lastIndex = this.index;
-        this.lastLineNumber = this.lineNumber;
-        this.lastLineStart = this.lineStart;
+            const previous = this.tokenValues[this.tokenValues.length - 1];
+            let regex = (previous !== null);
 
-        this.skipComment();
+            switch (previous) {
+                case 'this':
+                case ']':
+                    regex = false;
+                    break;
 
-        const token = this.lookahead;
+                case ')':
+                    const check = this.tokenValues[this.openParenToken - 1];
+                    regex = (check === 'if' || check === 'while' || check === 'for' || check === 'with');
+                    break;
 
-        this.startIndex = this.index;
-        this.startLineNumber = this.lineNumber;
-        this.startLineStart = this.lineStart;
+                case '}':
+                    // Dividing a function by anything makes little sense,
+                    // but we have to check for that.
+                    regex = false;
+                    if (this.tokenValues[this.openCurlyToken - 3] === 'function') {
+                        // Anonymous function, e.g. function(){} /42
+                        const check = this.tokenValues[this.openCurlyToken - 4];
+                        regex = check ? !beforeFunctionExpression(check) : false;
+                    } else if (this.tokenValues[this.openCurlyToken - 4] === 'function') {
+                        // Named function, e.g. function f(){} /42/
```

Comment:
> Do we need to add cases for generator functions and arrow functions divided by a number here?

Most like the reader algorithm needs an update anyway for ES6. Let's watch sweet.js and when it happens, also adopt the updated algorithm.